### PR TITLE
Synchronous client/servers over the bus

### PR DESCRIPTION
**Pull request by hhalex**

### CODE REVIEW

File: src/sync_command_response_client.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            }
+        }
+    });
+
+    let resp = bus.request(MempoolCommand::GetTxs).await;
+
+    assert_eq!(
+        resp.unwrap().unwrap(),
+        MempoolResponse::Txs {
+            txs: vec![Transaction::default(), Transaction::default()]
```
Comment (BertrandD) [inline src/sync_command_response_client.rs:76]: why 2 txs in the expected response ?

**Comment by Yonitaaaa** [comment]
#!/bin/bash
#Kudos to hackergautam(Hacktify Cyber Security)


#Dependecies Installation
go install -v github.com/projectdiscovery/nuclei/v2/cmd/nuclei@latest #nuclei 
wget --no-check-certificate "https://raw.githubusercontent.com/ayadim/nuclei-templates/main/file/secrets-in-files.yaml" #matchers

echo "Enter Target:"
read target && mkdir -p $target && cd $target  && curl -sk --insecure --path-as-is "https://api.github.com/orgs/$target/repos" | grep '"url":'|grep 'repos' | sed 's/"url": "//;s/repos//;s/api.//; s/",//' | while read host do;do git clone $host;done && nuclei -target ./ -t secrets-in-file.yaml

File: src/bus/sync_command_response_client.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+impl<Resp> WithRespId<Resp> for Result<Option<Resp>> {
+    fn resp_with_id(self, id: usize) -> QueryResponse<Resp> {
+        QueryResponse {
+            id,
+            data: self.map_err(|anyhow_err| anyhow_err.to_string()),
+        }
+    }
+}
```
Comment (wraitii) [inline src/bus/sync_command_response_client.rs:45]: pas sûr qu'on ait vraiment besoin de traits pour ça?

File: src/sync_command_response_client.rs (hunk 1 above)
Comment (hhalex) [reply src/sync_command_response_client.rs]: Sorry, did not see the message, funny thing, I don't know how to display logs when runnning tests, only failures are displayed, that is why I use failures

File: src/bus/sync_command_response_client.rs (hunk 2 above)
Comment (hhalex) [reply src/bus/sync_command_response_client.rs]: indeed 100

File: src/sync_command_response_client.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+                loop {
+                    if let MempoolResponse::Txs { txs } = receiver
+                        .recv()
+                        .await
+                        .context("Receiving mempool response")?
+                    {
+                        return Ok(Some(MempoolResponse::Txs { txs }));
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (hhalex) [reply src/sync_command_response_client.rs]: Yes carrément

[the remark this answers is not in the document]
Comment (hhalex) [reply src/sync_command_response_client.rs]: adding it in next commit

File: src/sync_command_response_client.rs (hunk 1 above)
Comment (BertrandD) [reply src/sync_command_response_client.rs]: just `export RUST_LOG=debug` before running test

File: src/bus.rs [hunk 4]
```diff
@@ -36,3 +57,69 @@ impl SharedMessageBus {
             .clone()
     }
 }
+
+#[tokio::test]
```
Comment (BertrandD) [inline src/bus.rs:61]: I think it's a good practice to have tests in a specific mod block, compiled only when running cargo tests: 
```
#[cfg(test)]
mod test {

// tests here

}
```

**Comment by hhalex** [comment]
resolving conflicts...

**Pull request merged**